src/passes/newToDeploy.ts

```diff
@@ -0,0 +1,215 @@
+        parent,
+      );
+      salt = createCallToFunction(narrowStub, [bytes32Salt], ast, parent);
+      ast.replaceNode(bytes32Salt, salt, parent);
+      ast.registerImport(salt, 'warplib.maths.utils', 'narrow_safe');
+    } else {
+      throw new TranspileFailedError(
+        `Contract New Expression has an unexpected parent. Expected FunctionCall or FunctionCallOptions instead ${
+          parent !== undefined ? printNode(parent) : 'undefined'
+        } was found`,
+      );
+    }
+
+    const placeHolderIdentifier = createIdentifier(placeholder, ast);
+    const deployCall = this.createDeploySysCall(
+      newFuncCall,
+      node.vTypeName,
+      placeHolderIdentifier,
+      salt,
+      ast,
+    );
+    ast.replaceNode(newFuncCall, deployCall);
+
+    this.visitExpression(node, ast);
+  }
+
+  private createDeploySysCall(
+    node: FunctionCall,
+    typeName: TypeName,
+    placeHolderIdentifier: Identifier,
+    salt: Expression,
+    ast: AST,
+  ): FunctionCall {
+    const deployStub = createCairoFunctionStub(
+      'deploy',
+      [
+        ['class_hash', createAddressTypeName(false, ast)],
+        ['contract_address_salt', createBytesNTypeName(30, ast)],
+        ['constructor_calldata', createBytesTypeName(ast), DataLocation.CallData],
+      ],
+      [['contract_address', cloneASTNode(typeName, ast)]],
+      ['syscall_ptr'],
+      ast,
+      node,
+      { acceptsUnpackedStructArray: true },
+    );
+    ast.registerImport(node, 'starkware.starknet.common.syscalls', 'deploy');
+
+    const encodedArguments = ast.getUtilFuncGen(node).utils.encodeAsFelt.gen(node);
+    return createCallToFunction(
+      deployStub,
+      [placeHolderIdentifier, salt, encodedArguments],
+      ast,
+      node,
+    );
+  }
+
+  private createPlaceHolder(
+    sourceUnit: SourceUnit,
+    declaredContract: ContractDefinition,
+    ast: AST,
+  ): VariableDeclaration {
+    const declaredContractSourceUnit = declaredContract.getClosestParentByType(SourceUnit);
+    assert(declaredContractSourceUnit !== undefined);
+
+    const declaredContractFullPath = declaredContractSourceUnit.absolutePath.split(
+      new RegExp('/+|\\\\+'),
+    );
+    const cairoPath = declaredContractSourceUnit.absolutePath
+      .slice(0, -'.sol'.length)
+      .concat('.cairo');
+
+    const fileName =
+      declaredContractFullPath[declaredContractFullPath.length - 1].split(CONTRACT_INFIX)[0];
+    const contractName = declaredContract.name;
+
+    const fullPath = declaredContractFullPath.slice(0, -1).join('_');
+    const varPrefix = `${fullPath}_${fileName}_${contractName}`;
+    const hash = hashFilename(cairoPath);
+    const varName = `${varPrefix}_${hash}`;
+
+    const varDecl = new VariableDeclaration(
+      ast.reserveId(),
+      '',
+      true,
+      false,
+      varName,
+      sourceUnit.id,
+      false,
+      DataLocation.Default,
+      StateVariableVisibility.Internal,
+      Mutability.Constant,
+      'uint160',
```

Comment:
@DomHenderson How do we feel about this being 160bits?

Reply:
I marked it as 160 bits thought in Cairo it can be bigger than that, it would never be an Uint256 (< 251 bits). Also it has the same size as an address. Perhaps making it an address type instead :thinking:

Reply:
Address might make the most sense I think. Not keen on it being 160 since with nethersolc there's no tie between the class hash and 160bits

Reply:
Could we have this as an Address and then we can confirm in the meeting on Wednesday